### ISSUE TITLE
[HOLD for payment 2024-08-02] [HOLD for payment 2024-08-01] Xero - App crashes when connecting to Xero

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#46039]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.11-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Sign in with a new expensifail account
2. Create a new workspace
3. Navigate to workspace settings >More features> Accounting
4. Enable accounting
5. Go to the Accounting settings page and click on Connect to Xero
6. Enable 2FA in the account and continue
7. The Xero site will be opened in a new tab
8. Enter the Xero credentials to log in and complete the connection
9. Go back to the staging app
10. Navigate away and back to the Accounting settings page

## Expected Result:
Xero connection is established and the accounting page opens properly

## Actual Result:
The app is stuck in establishing a connection and the app crashes when navigating away and back to the accountings page

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/fb4f1863-54a7-489b-a9ab-e96a27b0cfec

[logs (2).txt](https://github.com/user-attachments/files/16350945/logs.2.txt)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by francoisl** [comment]
I can repro the crash, it's most likely a NewDot issue and not backend.

**Comment by francoisl** [comment]
The crash only happens if you don't initially have 2FA enabled before connecting. The crash appears to be because we try to load the page at `/settings/workspaces/-1/accounting`, with an invalid workspace ID `-1`.


- 2FA enabled, no crash (the initial gets "stuck" on the step "Waiting for imported data to load", but that's a separate issue)

https://github.com/user-attachments/assets/ee38e862-a653-4958-bdd4-d4b9b1f39eec

--- 

- 2FA not enabled before connecting, app crash


https://github.com/user-attachments/assets/196bfafe-ca73-4871-b45f-13a8f65449bd

The console points to [this line](https://github.com/Expensify/App/blob/c5acf354eb6b443979ff23185fffca46a7438242/src/pages/workspace/accounting/PolicyAccountingPage.tsx#L147)

```
console.js:288 TypeError: Cannot read properties of undefined (reading 'id')
    at ce (PolicyAccountingPage.tsx:147:110)
```

But I don't think that's related, the last PR where that was modified has been on production for multiple days. Still trying to reliably repro in dev.

**Comment by francoisl** [comment]
Ok I think I see better what's happened. Right after enabled 2FA, on [these lines](https://github.com/Expensify/App/blob/c5acf354eb6b443979ff23185fffca46a7438242/src/libs/actions/Session/index.ts#L953-L954) - we clear the `policy_XXX` key from Onyx. Then inside the callback, we're supposed to call `openApp()`, but because of some errors we don't, and then when the user gets redirected back to NewDot, the workspace data doesn't exist, which causes the crash.

I see 20+ JS errors after enabling 2FA in the console, my guess is that clearing the Onyx cache causes some rerendering that relies on some keys that no longer exist. The first one is:

```
Unexpected text node: . A text node cannot be a child of a <View>. 
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at div
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at div
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at div
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at Section (webpack-internal:///./src/components/Section/index.tsx:39:68)
    at FeatureList (webpack-internal:///./src/components/FeatureList.tsx:25:68)
    at div
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at div
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at div
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/View/index.js:52:25)
    at eval (webpack-internal:///./node_modules/react-native-web/dist/exports/ScrollView/ScrollViewBase.js:68:24)
    at ScrollView (webpack-internal:///./node_modules/react-native-web/dist/exports/ScrollView/index.js:66:86)
    at ScrollView
    at ScrollView (webpack-internal:///./src/components/ScrollView.tsx:19:68)
```

cc @rushatgabhane if you have any idea as it's from https://github.com/Expensify/App/pull/45702. I don't think we want to revert again though, because the app doesn't work anyway without 2FA.

**Comment by rushatgabhane** [comment]
taking a look

**Comment by rushatgabhane** [comment]
@francoisl i think when we are connecting to Xero, we don't need to clear everything 😄 
Let me try it out

**Comment by francoisl** [comment]
Sounds good, let me know if it works. Alternatively, I tried to preserve active workspace data [like this](https://github.com/Expensify/App/pull/46065/files), so far it seems to fix the issue.

**Comment by rushatgabhane** [comment]
@francoisl it works. Thank you for investigating this, I'll raise a PR asap

**Comment by rushatgabhane** [comment]
> I tried to preserve active workspace data [like this](https://github.com/Expensify/App/pull/46065/files), so far it seems to fix the issue.

it is still a problem because a lot of data is missing

<img width="947" alt="image" src="https://github.com/user-attachments/assets/4ef224bd-e1eb-4252-b3d8-11912977ac44">

**Comment by francoisl** [comment]
Ah yeah, fair. Just an idea I was testing anyway. If you think we can skip clearing anything then it's probably better.

**Comment by rushatgabhane** [comment]
@francoisl we don't need Onyx clear at all. Because openApp returns the updated data.





<img width="1512" alt="image" src="https://github.com/user-attachments/assets/a61c728c-32f9-40e5-baf7-5e39ea677341">

**Comment by rushatgabhane** [comment]
But all keys with the method `mergeCollection` do not merge.
eg: policy, report, reportActions

<img width="781" alt="image" src="https://github.com/user-attachments/assets/4d7a3411-f975-4704-bf02-741178833825">

**Comment by rushatgabhane** [comment]
The root cause - the onyx keys with the method `mergeCollection` returned by `OpenApp` command do not merge. Making the app unusable until you clear everything.

**Comment by francoisl** [comment]
Are you saying this is an Onyx bug then?

**Comment by rushatgabhane** [comment]
No, I think Onyx.mergeCollection is not being invoked. 




https://github.com/user-attachments/assets/44f0f4f1-9953-4cb9-83fa-5a55abeeb816

**Comment by rushatgabhane** [comment]
When a signIn is done, two commands are called 
1. OpenApp
2. AuthenticatePusher

When 2FA for the admin flow is complete, only OpenApp is called.

**Comment by francoisl** [comment]
Could be related yeah, but still, the response from `OpenApp` has all the right data that should be directly be applied into Onyx.

Could it be some sort of race condition between clearing data and setting it back? After enabling 2FA, I don't even see non-collection things like `currencyList`, NVP keys, `country`, etc that `OpenApp` returned in Onyx.
<img width="1512" alt="image" src="https://github.com/user-attachments/assets/7424d14f-e125-4140-84bd-5581ca0aac13">

**Comment by rushatgabhane** [comment]
> Could it be some sort of race condition between clearing data and setting it back?

I removed Onyx.clear entirely

**Comment by francoisl** [comment]
Hm ok this issue is confusing 😅 

So removing the `Onyx.clear` fixes the bug described above in this issue (i.e. you try to connect a workspace to Xero when you don't have 2FA enabled yet); but it doesn't solve the issue where you sign in to an account that already has a Xero workspace and we force you to enable 2FA — is that right?

**Comment by rushatgabhane** [comment]
That is correct!!

**Comment by francoisl** [comment]
Ok then let's go with that for now then? At least it fixes the crash from this issue. And we can keep looking into the remaining issue later.

**Comment by francoisl** [comment]
Feel free to send a PR if you already have a WIP branch somewhere.

**Comment by rushatgabhane** [comment]
https://github.com/Expensify/App/pull/46087

**Comment by rushatgabhane** [comment]
Tests well @francoisl

**Comment by francoisl** [comment]
Thanks! I have to go for a bit but I'll be back later to review.

**Comment by francoisl** [comment]
Tested on staging web, 👍

**Comment by rushatgabhane** [comment]
This is a regression fix. No payment required

**Comment by rushatgabhane** [comment]
@francoisl good to close
